dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -177,52 +176,36 @@ internal KernelFunction(string name, string? pluginName, string description, IRe
             var invokingEventArgs = kernel.OnFunctionInvoking(this, arguments);
 #pragma warning restore CS0618 // Events are deprecated
 
-            // Invoke pre-invocation filter. If it requests cancellation, throw.
-            var invokingContext = kernel.OnFunctionInvokingFilter(this, arguments);
-
             if (invokingEventArgs?.Cancel is true)
             {
                 throw new OperationCanceledException($"A {nameof(Kernel)}.{nameof(Kernel.FunctionInvoking)} event handler requested cancellation before function invocation.");
             }
 
-            if (invokingContext?.Cancel is true)
+            var invocationContext = await kernel.OnFunctionInvocationAsync(this, arguments, functionResult, async (context) =>
             {
-                throw new OperationCanceledException("A function filter requested cancellation before function invocation.");
-            }
+                // Invoking the function and updating context with result.
+                context.Result = functionResult = await this.InvokeCoreAsync(kernel, context.Arguments, cancellationToken).ConfigureAwait(false);
+            }).ConfigureAwait(false);
 
-            // Invoke the function.
-            functionResult = await this.InvokeCoreAsync(kernel, arguments, cancellationToken).ConfigureAwait(false);
+            // Apply any changes from the function filters context to final result.
+            functionResult = new FunctionResult(invocationContext.Result);
```

Comment:
Our primary goal here is to override `functionResult` values from `invocationContext.Result`. We can do it either by assigning each property one-by-one (e.g. `functionResult.Function = invocationContext.Result.Function`), or we can use new constructor that I've added, which accepts source `FunctionResult` to copy data from:
![image](https://github.com/microsoft/semantic-kernel/assets/13853051/032ae506-bc4f-4435-bf56-2c829fb2b2e9)

The benefit of using this constructor is when we will add new property to `FunctionResult`, we will need to update just this constructor with that property, and all places like this will automatically copy new properties without doing modifications in code where `functionResult` properties have to be overridden. There are multiple places with such logic, including function filter, auto function invocation filter and streaming/non-streaming scenarios for each filter.

Reply:
Yes, I get that.

My question is about the copy we create, can we simply do `functionResult = invocationContext.Result;`

Reply:
Oh yes, I completely forgot about that option 😄 
I don't think that copying provides some benefits here, so re-assigning should be good. I will try to do that and see if everything still works as expected.

Reply:
@markwallace-microsoft I decided to leave it as it is and use copying, as a safer option. When we do `functionResult = invocationContext.Result;`, we just change the reference of `functionResult` variable. Taking into account that `context.Result` can be overridden in filter, we may end up with situation when `kernel.InvokeAsync(function)` will return the same object which was set in function filter. It's not necessarily bad approach, but I feel like functions should be completely isolated from filters and return their own instances rather than instances that were created in filters. Copying should not affect performance, and we will have more control in the future in case we decide to copy not all properties, but just some of them. Let me know if that makes sense.

Reply:
Makes sense, I updated the code.